sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -687,9 +690,94 @@ def get_cmis_media_lanes_mask(self, api, appl, lport, subport):
 
         return media_lanes_mask
 
-    def is_appl_reconfigure_required(self, api, app_new):
+    def clear_decomm_pending(self, lport):
         """
-	   Reset app code if non default app code needs to configured 
+        Clear the decommission pending status for the entire physical port this logical port belongs to.
+
+        Args:
+            lport:
+                String, logical port name
+        """
+        self.decomm_pending_dict.pop(self.port_dict.get(lport, {}).get('index'), None)
+
+    def set_decomm_pending(self, lport):
+        """
+        Set the decommission pending status.
+
+        Args:
+            lport:
+                String, logical port name
+        """
+        physical_port_idx = self.port_dict[lport]['index']
+        if physical_port_idx in self.decomm_pending_dict:
+            # only one logical port can be the lead logical port doing the
+            # decommission state machine.
+            return
+        self.decomm_pending_dict[physical_port_idx] = lport
+        self.log_notice("{}: DECOMMISSION: setting decomm_pending for physical port "
+                        "{}".format(lport, physical_port_idx))
+
+    def is_decomm_lead_lport(self, lport):
+        """
+        Check if this is the lead logical port doing the decommission state machine.
+
+        Args:
+            lport:
+                String, logical port name
+        Returns:
+            Boolean, True if decommission pending, False otherwise
+        """
+        return self.decomm_pending_dict.get(self.port_dict[lport]['index']) == lport
+
+    def is_decomm_pending(self, lport):
+        """
+        Get the decommission pending status for the physical port the given logical port belongs to.
+
+        Args:
+            lport:
+                String, logical port name
+        Returns:
+            Boolean, True if decommission pending, False otherwise
+        """
+        return self.port_dict[lport]['index'] in self.decomm_pending_dict
+
+    def is_decomm_failed(self, lport):
+        """
+        Get the decommission failed status for the physical port the given logical port belongs to.
+
+        Args:
+            lport:
+                String, logical port name
+        Returns:
+            Boolean, True if decommission failed, False otherwise
+        """
+        physical_port_idx = self.port_dict[lport]['index']
+        lead_logical_port = self.decomm_pending_dict.get(physical_port_idx)
+        if lead_logical_port is None:
+            return False
+        return (
+            get_cmis_state_from_state_db(
+                lead_logical_port,
+                self.xcvr_table_helper.get_status_sw_tbl(
+                    self.get_asic_id(lead_logical_port)
+                )
+            )
+            == CMIS_STATE_FAILED
+        )
+
+    def is_decommission_required(self, api, app_new):
```

Comment:
@AnoopKamath  This API will work only in cases where the module is operating with same application across all the datapath. Is that correct understanding? If there is a mix of application then?

Reply:
@prgeor : Your understanding is correct. This PR addresses DP INIT issues during decommission by incorporating delays recommended by modules, and prevents configuration failures during decommissioning. @longhuan-cisco is working on the code changes to handle mixed applications and speed changes (https://github.com/sonic-net/sonic-buildimage/issues/23006) scenarios. Will send out the review soon.